### ISSUE TITLE
MySQL charset not restored upon reconnection

### ISSUE THREAD

**Issue by CrazyHackGUT** [alliedmodders/sourcemod#1173]
Sometimes we need load some plugins before all anothers for priority in forward calling.
Currently, SM doesn't provide any method for changing load priority.

**Comment by dvander** [comment]
Plugins should be agnostic of load order, otherwise, you could get circular dependencies.

**Comment by Headline** [comment]
We already should be handling this internally, what's the issue you're hitting specifically?

**Comment by CrazyHackGUT** [comment]
> what's the issue you're hitting specifically?

I have some plugins who works with MySQL. This plugins doesn't supports `utf8mb4`, so i wrote for myself usage small plugin who remove all special characters from nicknames.
When plugin is uploaded on server as `ClientNameFixer.smx`, some plugins (which are displayed first before my) builds query with special characters such as smilies or any another dumbest symbols who is not be saved without charset `utf8mb4`.
When i rename my plugin into `1.smx` and reload server, my plugin displayed as first in plugin list, and this issue is disappears.

**Comment by dvander** [comment]
There is fundamentally no way to solve this problem, since two plugins could ask to be "first". Your best bet is to fix the plugins not supporting utf8.

**Comment by CrazyHackGUT** [comment]
I will be happy to do it, seriously.
But... I have problem with MySQL driver.

[When MySQL driver losts a connection and reconnects](https://github.com/alliedmodders/sourcemod/blob/5611ec54a21c3045cc1680b954631c6ca049c768/extensions/mysql/mysql/MyDriver.cpp#L110-L114), it does not restore the previously set encoding. The only solution I found to deal with this is to set the encoding before each query, what creates freezes (`SQL_SetCharset` is synchoronous).
Not for all databases i have access to MySQL server configuration for changing default character set.

**Comment by dvander** [comment]
Can we fix this at the DBI level by detecting lost connections?On Feb 6, 2020 7:20 PM, Kruzya <notifications@github.com> wrote:I will be happy to do it, seriously.
But... I have problem with MySQL driver.
When MySQL driver losts a connection and reconnects, it does not restore the previously set encoding. The only solution I found to deal with this is to set the encoding before each query, what creates freezes (SQL_SetCharset is synchoronous).
Not for all databases i have access to MySQL server configuration for changing default character set.

**Comment by CrazyHackGUT** [comment]
> Can we fix this at the DBI level by detecting lost connections?

In [MySQL C API documentation about reconnecting](https://dev.mysql.com/doc/refman/5.7/en/c-api-auto-reconnect.html) for this cases recommended use `mysql_options()` and `MYSQL_INIT_COMMAND`. But this function should be called only before `mysql_connect()`. So i don't understand how exactly you are going to detect lost connection for applying charset after reconnection.

**Comment by CrazyHackGUT** [comment]
Maybe move this problem with MySQL to separate issue? If SM developers can fix it, i will be happy, and i remove my epic plugin-scratch with usernames.

**Comment by peace-maker** [comment]
SQL_SetCharset uses `mysql_set_character_set` which does exactly what you've described. It sets the desired charset for the connection in the mysql client library which gets reapplied in case of a reconnect automatically.

So you only need to call that native once in your `SQLConnectCallback` function after the first connection to the database and it will use the set charset for all future queries. No need to call it before every query.

See the bug which lead to introducing `SQL_SetCharset`.
https://bugs.alliedmods.net/show_bug.cgi?id=5786
https://github.com/alliedmodders/sourcemod/commit/69d8b1c30dbf1a12c8a1b572049427a11bbbe092

**Comment by CrazyHackGUT** [comment]
> It sets the desired charset for the connection in the mysql client library which gets reapplied in case of a reconnect automatically.

This is don't work. I tested very many times, but this is really don't work.
Users of my one plugin complain on hieroglyphs when execution `SQL_SetCharset()` before all queries is disabled (only after creating connection).

**Comment by wopox1337** [comment]
I confirm the presence of this problem. I ran into her recently

**Comment by vadrozh** [comment]
Really need to fix this

**Comment by asherkin** [comment]
> SQL_SetCharset uses `mysql_set_character_set` which does exactly what you've described. It sets the desired charset for the connection in the mysql client library which gets reapplied in case of a reconnect automatically.

Turns out it doesn't - it sets the current connection's internal charset, but doesn't set the charset to apply on reconnection.

![image](https://user-images.githubusercontent.com/865360/76153804-63f64b80-60c9-11ea-9a75-45a06da0b0dd.png)

**Comment by peace-maker** [comment]
That sounds like a regression in the mysql client library? It's been fine when we were at mysql 5.0.

**Comment by asherkin** [comment]
> That sounds like a regression in the mysql client library? It's been fine when we were at mysql 5.0.

Not sure, the [upstream docs](https://dev.mysql.com/doc/refman/5.6/en/c-api-auto-reconnect.html) seem fairly explicit about it:

> Automatic reconnection can be convenient because you need not implement your own reconnect code, but if a reconnection does occur, several aspects of the connection state are **reset on the server side** and your application will not be notified.
> 
> The connection-related state is affected as follows:
>  * [...]
>  * Session system variables are reinitialized to the values of the corresponding global system variables, **including system variables that are set implicitly by statements such as SET NAMES**.
>  * [...]

(Emphasis mine.)

**Comment by peace-maker** [comment]
Looking at the source (just some mysql version that popped up first):
in [mysql_reconnect](https://github.com/charliesome/mysql-5.5.36/blob/master/sql-common/client.c#L3605):
```c
if (mysql_set_character_set(&tmp_mysql, mysql->charset->csname))
```
it uses `mysql->charset` to reappyl the charset.

[mysql_set_character_set](https://github.com/charliesome/mysql-5.5.36/blob/master/sql-common/client.c#L4272) does
```c
mysql->charset= cs;
```

Are you sure `mysql->options.charset_name` is used during reconnection?

**Comment by asherkin** [comment]
Hmm, didn’t realise it explicitly did the full charset set on after reconnecting - yeah, that should all be working fine then.

**Comment by CrazyHackGUT** [comment]
> It's been fine when we were at mysql 5.0.

No, it works absolutely identital bad on 5.0 too (SM 1.8 and lower).

**Comment by asherkin** [comment]
> > It's been fine when we were at mysql 5.0.
> 
> No, it works absolutely identital bad on 5.0 too (SM 1.8 and lower).

It would be really helpful if you could provide step-by-step instructions for reliably reproducing the issue with a minimal plugin and config.

**Comment by CrazyHackGUT** [comment]
> It would be really helpful if you could provide step-by-step instructions for reliably reproducing the issue with a minimal plugin and config.

Sorry, but i can't reproduce it manually. It is not reproduced specifically for me. I can only give link on plugin, where this is reproduces very often and give some crash information (from Accelerator, yes).

**Comment by psychonic** [comment]
Closing as stale since there isn't a way to reliably reproduce and it's likely a bug in the mysql client itself since it's documented to handled this automatically.
